core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -171,33 +178,33 @@ public final Object visitClearCommand(InvocationContext ctx, ClearCommand comman
       getCommand.setTopologyId(topologyId);
       getCommand.setWrite(isWrite);
 
-      return rpcManager.invokeRemotelyAsync(info.readOwners(), getCommand, getStaggeredOptions(info.readOwners().size())).thenAccept(responses -> {
-         for (Response r : responses.values()) {
-            if (r instanceof SuccessfulResponse) {
-               SuccessfulResponse response = (SuccessfulResponse) r;
-               Object responseValue = response.getResponseValue();
-               if (responseValue == null) {
-                  if (rvrl != null) {
-                     rvrl.remoteValueNotFound(key);
-                  }
-                  wrapRemoteEntry(ctx, key, NullCacheEntry.getInstance(), isWrite);
-                  return;
-               }
-               InternalCacheEntry ice = ((InternalCacheValue) responseValue).toInternalCacheEntry(key);
-               if (rvrl != null) {
-                  rvrl.remoteValueFound(ice);
-               }
-               wrapRemoteEntry(ctx, key, ice, isWrite);
-               return;
-            }
-         }
-         throw handleMissingSuccessfulResponse(responses);
-      });
-   }
-
-   protected static CacheException handleMissingSuccessfulResponse(Map<Address, Response> responses) {
+      return rpcManager.invokeCommandStaggered(info.readOwners(), getCommand, new RemoteGetResponseCollector(),
+                                               rpcManager.getSyncRpcOptions())
+                       .thenAccept(r -> {
+                          if (r instanceof SuccessfulResponse) {
+                             SuccessfulResponse response = (SuccessfulResponse) r;
+                             Object responseValue = response.getResponseValue();
+                             if (responseValue == null) {
+                                if (rvrl != null) {
+                                   rvrl.remoteValueNotFound(key);
+                                }
+                                wrapRemoteEntry(ctx, key, NullCacheEntry.getInstance(), isWrite);
+                                return;
+                             }
+                             InternalCacheEntry ice = ((InternalCacheValue) responseValue).toInternalCacheEntry(key);
+                             if (rvrl != null) {
+                                rvrl.remoteValueFound(ice);
+                             }
+                             wrapRemoteEntry(ctx, key, ice, isWrite);
+                             return;
+                          }
+                          throw handleMissingSuccessfulResponse(r);
```

Comment:
I don't see a need for an extra check, can you elaborate?

As the callback is added to the request I think it should be pretty clear that it can't see responses before the request is completed. In principle I'm open to the idea of using more customized collectors, but I don't see the case for it here.